autogpt_platform/autogpt_libs/autogpt_libs/utils/cache.py

```diff
@@ -200,48 +291,98 @@ def sync_wrapper(*args: P.args, **kwargs: P.kwargs):
                                     return result
 
+                # Slow path: acquire lock for cache miss/expiry
+                with cache_lock:
+                    # Double-check: another thread might have populated cache
+                    if shared_cache:
+                        try:
+                            # Use GET directly (saves 1 round trip)
+                            cached_bytes = redis_client.get(redis_key)
+                            if cached_bytes is not None and isinstance(
+                                cached_bytes, bytes
+                            ):
+                                return pickle.loads(cached_bytes)
+                        except Exception as e:
+                            logger.error(
+                                f"Redis error during double-check for {target_func.__name__}: {e}"
+                            )
+                            # Continue to execute function
+                    else:
+                        if key in cache_storage:
+                            if ttl_seconds is None:
+                                return cache_storage[key]
+                            else:
+                                cached_data = cache_storage[key]
+                                if isinstance(cached_data, tuple):
+                                    result, timestamp = cached_data
+                                    if current_time - timestamp < ttl_seconds:
+                                        return result
+
                     # Cache miss - execute function
                     logger.debug(f"Cache miss for {target_func.__name__}")
                     result = target_func(*args, **kwargs)
 
                     # Store result
-                    if ttl_seconds is None:
-                        cache_storage[key] = result
+                    if shared_cache:
+                        try:
+                            pickled_result = pickle.dumps(
+                                result, protocol=pickle.HIGHEST_PROTOCOL
+                            )
+                            if ttl_seconds is None:
+                                redis_client.set(redis_key, pickled_result)
+                            else:
+                                redis_client.setex(
+                                    redis_key, ttl_seconds, pickled_result
+                                )
+                        except Exception as e:
+                            logger.error(
+                                f"Redis error storing cache for {target_func.__name__}: {e}"
+                            )
+                            # Continue without caching
                     else:
-                        cache_storage[key] = (result, current_time)
+                        if ttl_seconds is None:
+                            cache_storage[key] = result
+                        else:
+                            cache_storage[key] = (result, current_time)
```

Comment:
This makes me wonder, if we do this, how do we differentiate which entry has ttl and not? because the result can be also a tuple too and we won't have a way to know whether the second entry is part of the result or the expiry time.